lib/orm.php

```diff
@@ -2541,4 +2589,124 @@ public function __unset( $key ) {
 	public function __isset( $key ) {
 		return $this->offsetExists( $key );
 	}
+
+	/**
+	 * Whether this site has persistent caching implemented.
+	 *
+	 * @return bool
+	 */
+	private static function has_persistent_cache() {
+		if ( self::$has_persistent_cache === false || self::$has_persistent_cache === true ) {
+			return self::$has_persistent_cache;
+		}
+
+		// Check if the site is using an external object cache.
+		if ( \function_exists( 'wp_using_ext_object_cache' )
+			&& \wp_using_ext_object_cache()
+		) {
+			self::$has_persistent_cache = true;
+			return true;
+		}
+
+		// Defensive coding - cover all bases.
+		// This will prevent errors below if WP_CONTENT_DIR is not defined.
+		if ( ! defined( 'WP_CONTENT_DIR' ) ) {
+			return false;
+		}
+
+		// Check if the site is using a persistent drop-in (object-cache.php).
+		// This _should_ be covered by the `wp_using_ext_object_cache()` check,
+		// but sometimes isn't depending on the timing of the cache.
+		if ( \file_exists( WP_CONTENT_DIR . '/object-cache.php' ) ) {
+			self::$has_persistent_cache = true;
+			return true;
+		}
+
+		// No persistent caching was detected.
+		self::$has_persistent_cache = false;
+		return false;
+	}
+
+	/**
+	 * Get a cache.
+	 *
+	 * @param string $key      Cache key.
+	 * @param mixed  $group_id Cache group ID.
+	 *
+	 * @return mixed The cache value.
+	 */
+	public static function get_cache( $key, $group_id = '' ) {
+		// Bail early if the site doesnt have persistent caching.
+		if ( ! self::has_persistent_cache() ) {
+			return false;
+		}
+		$cache_group = self::get_cache_group( $group_id );
+		return \wp_cache_get( $key, $cache_group );
+	}
+
+	/**
+	 * Set a cache.
+	 *
+	 * @param string $key      Cache key.
+	 * @param mixed  $value    Cache value.
+	 * @param mixed  $group_id Cache group ID.
+	 *
+	 * @return void
+	 */
+	private static function set_cache( $key, $value, $group_id = '' ) {
+		// Bail early if the site doesnt have persistent caching.
+		if ( ! self::has_persistent_cache() ) {
+			return;
+		}
+		/**
+		 * Set the cache, with a timeout of 10 minutes.
+		 *
+		 * The 10-minute is long enough to significantly improve performance,
+		 * but short enough to not cause issues with stale data.
+		 *
+		 * Individual cache-groups are purged when a record is updated or deleted.
+		 */
+		\wp_cache_set( $key, $value, self::get_cache_group( $group_id ), ( 10 * MINUTE_IN_SECONDS ) );
```

Comment:
Instead of doing this, by do correct cache invalidation. there are lots of actions in core to let you do this.

Reply:
We're already invalidating cache in this PR, but we want to take the better-safe-than-sorry route here and roll out the enhancement with a short TTL as a first iteration. 

Once that's out and battle tested in the wild, we can then consider to increase the TTL in later iterations, for greater performance impact.

Reply:
We can (and should, probably) make the TTL filterable though, to allow customization.

Reply:
If you are doing cache invalidation, you should not need to only cache for 10 minutes. Just let the invalidation handle itself. A 10 minute cache is not specially useful. For most high traffic sites, they would be behind a CDN and traffic might not hit the origin for a url every 10 minutes. Object caches like this are useful for sites with lots of traffic and content. So a site with 100k posts, might only get 1-5 hits a day on an article from 5 years ago. But if google indexes the site, that number of db queries can take a site down.

Reply:
OK, now I'm confused, can you please elaborate? Because I'm probably missing context. 

I thought what prompted all this is the need for putting queries behind cache to avoid repeated db load within a short period of time. Copying from [your original post](https://github.com/Yoast/wordpress-seo/issues/17005):
> but for sites under heavily load or under a DDOS attack, 3 queries a page should take the database server and website down.

For that reason, we came up with that staggered strategy I mention above. To take care of such cases (and then expand our feature to a longer caching period) in a safe and contained way. 

We have to keep in mind that what we're trying to cache here has serious SEO implications and any inconsistency brought by cache has the potential to hurt websites in that regard.

> So a site with 100k posts, might only get 1-5 hits a day on an article from 5 years ago.
If so, are we caching here just to make those 1-5 queries a day into 0?

If that's the case, we should also re-consider our whole current implementation. Because they way we do it now, cache invalidation happens with every insert/update/delete to the table in play with each query, which has the potential to happen rather frequently (with every new post, with every updated post, with every post **visit** (if SEO optimization hasn't been performed in that site), with every user addition, with every term addition, etc. etc.)

@aristath Do you have any thoughts on all that?

Reply:
The cache key should be generated from the object and id. Cache invalidation should done on that cache key when only that object is changed. Not every update to the database.